### PR TITLE
Allow conditions inside loops to preserve or narrow int range

### LINKED ISSUE #7555: Incorrect type checking (0|positive-int)
https://psalm.dev/r/0b7a15278a

### LINKED ISSUE #8020: Int range not updated when in loop
https://psalm.dev/r/65c57ad265

I figured this would work since the variable is redefined on each loop iteration.

### LINKED ISSUE #8865: Psalm fails to narrow type to positive int inside loop
Psalm fails to infer that a variable is a positive int when the condition occurs inside a loop.

https://psalm.dev/r/155977bb20

### PR DESCRIPTION
**Pull request by theodorejb**
Fixes #7555, fixes #8020, fixes #8865

I lack familiarity with the assertion reconciler code, so I'm not sure if there is something the continue if `$inside_loop` checks were necessary for. But at least the current tests still seem to pass without them, and this fixes issues where conditions inside a loop failed to preserve or narrow the range of an int as expected.

### CODE REVIEW

File: tests/IntRangeTest.php [hunk 1]
```diff
@@ -457,6 +456,35 @@ function takesInt(int $i): void{
...
+                        }
+                    }
+                    takesNonNegativeInt($sum);
+
+                    /** @psalm-param int<0, max> $i */
+                    function takesNonNegativeInt(int $i): void{
+                        return;
+                    }',
+            ],
             'SKIPPED-wrongLoopAssertion' => [
```
Comment (orklah) [inline tests/IntRangeTest.php:488]: Can you try to unskip that one too?

File: src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php [hunk 2]
```diff
@@ -123,10 +123,6 @@ public static function isContainedBy(
         }
 
         foreach ($input_type_part->type_params as $i => $input_param) {
-            if ($i > 1) {
```
Comment (orklah) [inline src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126]: why did you remove that one?

Comment (theodorejb) [reply src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126]: It causes the following error:
> TypeDoesNotContainType - src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126:17 - Type int<0, 1> for $i is never >1

File: tests/IntRangeTest.php (hunk 1 above)
Comment (theodorejb) [reply tests/IntRangeTest.php:488]: I'm not sure what that one is trying to test, but it causes the following error:
> Psalm\Exception\CodeException: Trace - src\somefile.php:8:51 - $i: int<0, max>

File: src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php (hunk 2 above)
Comment (orklah) [reply src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126]: yep but that's wrong. This loop can go past one

File: tests/IntRangeTest.php (hunk 1 above)
Comment (orklah) [reply tests/IntRangeTest.php:488]: Strange that traces were left in there. Can you get rid of the trace comments to see if there is other errors?

File: src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php (hunk 2 above)
Comment (orklah) [reply src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126]: That must be one of the reasons I skipped tests and did not try analyzing assertions inside loops. Unfortunately, this is not ready

File: tests/IntRangeTest.php (hunk 1 above)
Comment (theodorejb) [reply tests/IntRangeTest.php]: There aren't any errors after removing the trace comments.

File: src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php (hunk 2 above)
Comment (theodorejb) [reply src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126]: What would cause the loop to go past one since the `type_params` property has a type of `array{Union, Union}`? And what is the `break` needed for that isn't covered by a test?

Comment (orklah) [reply src/Psalm/Internal/Type/Comparator/ArrayTypeComparator.php:126]: Oh yeah, I guess you're right. I thought it was just a performance thing but the loop can't go over 1 indeed.

Array count has improved a lot recently, maybe it has fixed things that were broken when I decided to skip processing assertions in loop

**Comment by orklah** [comment]
Thanks!
